### PR TITLE
(fix) Waveform scratch crossing loop boundaries

### PR DESCRIPTION
**Pull request by ronso0**
Previously, scratching with the waveform and crossing loop boundaries at slow rates caused an infinite wrap-around 'loop' even after the mouse was stopped (while still pressed).
Probably affects Spinny scratching, too, but I didn't test it.

PositionScratchController::process() now knows if the play position was wrapped around and calculates the true sample distance travelled.
_I notice this would also be helpful for and probably simplify #12951 hotcue_X_indicator control_

TL;DR
PositionScratchController didn't know the play position was wrapped around, so it
interpreted the raw (huge) sample delta as play pos move and calculated a huge rate, which was
then applied by the engine, resulting in a (true) long sample distance in the next run,
which caused another loop wrap-around, and so on...

### CODE REVIEW

**Comment by ronso0** [comment]
This is now cleaned up.
You may pick d4a0daa179 to add some debug output.

File: src/engine/controls/ratecontrol.cpp [hunk 1]
```diff
@@ -603,3 +616,18 @@ bool RateControl::isReverseButtonPressed() {
     }
     return false;
 }
+
+void RateControl::notifyWrapAround(mixxx::audio::FramePos triggerPos,
+        mixxx::audio::FramePos targetPos) {
+    // TODO FramePos only for easy validation.
+    // Check how to get kLegacyInvalidEnginePosition in order to use sample pos
+    // for the entire
```
Comment (JoergAtGithub) [inline src/engine/controls/ratecontrol.cpp:624]: Why do you want to switch from frame position to sample position?

**Review by JoergAtGithub — COMMENTED** [review]
I tested this on Windows and everything works as expected!

**Comment by Swiftb0y** [comment]
I remember mentioning this in an issue or zulip post sometime ago, but I can't find it anymore

**Comment by ronso0** [comment]
yeah, we touched this topic sometime, but I can't find it anymore either.

File: src/engine/controls/ratecontrol.cpp (hunk 1 above)
Comment (ronso0) [reply src/engine/controls/ratecontrol.cpp:624]: it's passed only to PositionScratchController and uses only sample positions for processing anyway, so I thought we'd better convert it early. No issue though, and getting kLegacyInvalidEnginePosition here requires more includes, so I can as well remove the TODO.

**Comment by ronso0** [comment]
> I tested this on Windows and everything works as expected!

An important test case, which failed with my previous attempts, is a fast throwback with a very short loop (< 1 beat), then hold the mouse button until scratching (inertia) stopped. This should perform exactly like a throwback with no loop.
And scracthing should also **sound** right, not just look reasonable on the waveforms.

**Comment by JoergAtGithub** [comment]
> > I tested this on Windows and everything works as expected!
> 
> An important test case, which failed with my previous attempts, is a fast throwback with a very short loop (< 1 beat), then hold the mouse button until scratching (inertia) stopped. This should perform exactly like a throwback with no loop. And scracthing should also **sound** right, not just look reasonable on the waveforms.

I tested this, and it worked correct!

**Comment by ronso0** [comment]
Great. So I remove the TODO and this is ready?

**Comment by ronso0** [comment]
Done.

**Comment by JoergAtGithub** [comment]
LGTM! Thank you!
